### PR TITLE
Add Proxy Auto-Configuration (PAC) support

### PR DESCRIPTION
**Pull request by ThinkChaos**
I've been following Passepartout's development for a while and would love to use the iOS client full-time, but the feature I'm currently missing is PAC support. So here's a PR!

### CODE REVIEW

File: TunnelKit/Sources/Protocols/OpenVPN/AppExtension/OpenVPNTunnelProvider+Configuration.swift [hunk 1]
```diff
@@ -181,6 +181,8 @@ extension OpenVPNTunnelProvider {
             
             static let httpsProxy = "HTTPSProxy"
             
+            static let proxyAutoConfURL = "proxyAutoConfURL"
```
Comment (keeshux) [inline TunnelKit/Sources/Protocols/OpenVPN/AppExtension/OpenVPNTunnelProvider+Configuration.swift]: Please replace this declaration with:

`static let proxyAutoConfigurationURL = "ProxyAutoConfigurationURL"`

which is more in accordance to (unsaid) key name conventions.

Needless to say, rename `proxyAutoConfigURL` to `proxyAutoConfigurationURL` everywhere else.

File: TunnelKit/Sources/Protocols/OpenVPN/Configuration.swift [hunk 2]
```diff
@@ -244,9 +244,12 @@ extension OpenVPN {
         /// The search domain.
         public var searchDomain: String?
 
+        /// The Proxy Auto-Configuration (PAC) url.
+        public var proxyAutoConfURL: URL?
```
Comment (keeshux) [inline TunnelKit/Sources/Protocols/OpenVPN/Configuration.swift]: Likewise, use `proxyAutoConfigurationURL`.

File: TunnelKit/Sources/Protocols/OpenVPN/ConfigurationParser.swift [hunk 3]
```diff
@@ -517,6 +518,17 @@ extension OpenVPN {
                     optSearchDomain = $0[1]
                 }
                 Regex.proxy.enumerateArguments(in: line) {
+                    if $0.count == 2 {
+                        let maybeURL = URL(string: $0[1])
+                        if maybeURL != nil {
```
Comment (keeshux) [inline TunnelKit/Sources/Protocols/OpenVPN/ConfigurationParser.swift]: Any reason against an `if let`? Or `guard let`, much better.

File: TunnelKit/Sources/Protocols/OpenVPN/AppExtension/OpenVPNTunnelProvider.swift [hunk 4]
```diff
@@ -665,18 +668,26 @@ extension OpenVPNTunnelProvider: OpenVPNSessionDelegate {
...
             log.info("Routing: Setting HTTP proxy \(httpProxy.address.maskedDescription):\(httpProxy.port)")
         }
+        if let pacURL = cfg.sessionConfiguration.proxyAutoConfURL ?? options.proxyAutoConfURL {
+            if proxySettings == nil {
+                proxySettings = NEProxySettings()
+            }
+            proxySettings!.proxyAutoConfigurationURL = pacURL
+            proxySettings!.autoProxyConfigurationEnabled = true
+            log.info("Routing: Setting PAC \(pacURL)")
+        }
```
Comment (keeshux) [inline TunnelKit/Sources/Protocols/OpenVPN/AppExtension/OpenVPNTunnelProvider.swift]: While I understand that the result is 100% equal, I like to cut unwrapped optionals to the bare minimum. Seeing an "!" automatically rings a bell and has me wonder "is it safe to use here?", even if this is an obviously safe place to use it. So I'd rather be "visually safe" too by using the "?" instead.

Could you therefore use `proxySettings?` vs `proxySettings!` here?

File: TunnelKit/Sources/Protocols/OpenVPN/ConfigurationParser.swift [hunk 5]
```diff
@@ -517,6 +518,17 @@ extension OpenVPN {
...
+                            optProxyAutoConfURL = maybeURL!
+                        }
+                        else {
+                            unsupportedError = ConfigurationError.malformed(option: "dhcp-option PROXY_AUTO_CONFIG_URL has malformed URL")
+                        }
+                        return
+                    }
+
```
Comment (keeshux) [inline TunnelKit/Sources/Protocols/OpenVPN/ConfigurationParser.swift]:
```
guard let maybeURL = URL(string: $0[1]) else {
    unsupportedError = ConfigurationError.malformed(option: "dhcp-option PROXY_AUTO_CONFIG_URL has malformed URL")
    return
}
optProxyAutoConfURL = maybeURL
```

**Comment by keeshux** [comment]
@ThinkChaos thank you very much, this is definitely the best way to have a forgotten feature in reasonable time. 😄

File: TunnelKit/Sources/Protocols/OpenVPN/ConfigurationParser.swift (hunk 3 above)
Comment (ThinkChaos) [reply TunnelKit/Sources/Protocols/OpenVPN/ConfigurationParser.swift]: Yeah XCode was buggy and when I debugged the test it was showing the url as `nil` so I wrote this (which didn't solve the XCode issue) but forgot to change it back!

File: TunnelKit/Sources/Protocols/OpenVPN/AppExtension/OpenVPNTunnelProvider.swift (hunk 4 above)
Comment (ThinkChaos) [reply TunnelKit/Sources/Protocols/OpenVPN/AppExtension/OpenVPNTunnelProvider.swift]: I like to use `!` in places where it's obvious it won't error/or add a comment explaining why it's ok and use it as an assert. I'll change it back to `?` to match your style.

**Comment by ThinkChaos** [comment]
Thank you for responding so fast. Looking forward to have this in the iOS Beta 😉

**Comment by keeshux** [comment]
No reason to wait further, code was solid and nearly harmless to existing codebase. New beta this week.

**Pull request merged**